tools/offline_log_viewer/data_directory_inspector.py

```diff
@@ -0,0 +1,899 @@
+                                                       # but lacks archival_metadata.snapshot
+      tx.snapshot                                      # has rm_stm's snapshot which is a bug
+      tx.snapshot.partial.1621.a8D9
+
+redpanda/                                              # another place to store inernal topics: controller & kvstore
+  controller/
+    0_1/                                               # has only one partition, no persisted stm's
+      1337-2-v1.log
+      1337-2-v1.base_index
+      snapshot
+  kvstore/                                             # kvstore is local non-replicated topic, different nodes have
+                                                       # different kvstore logs; it's used to keep aux info such as
+                                                       # raft's votes of the current node
+    0_1/
+      1337-2-v1.log
+      1337-2-v1.base_index
+      snapshot                                         # name collision, this isn't the raft snapshot
+
+cloud_storage_cache/                                   # SI cache
+  accesstime
+  f6ek42xx/                                            # 8 symbol hash
+    kafka/
+      topic1/                                          # name of the cached topic
+        12_5/                                          # partitions {id}_{revision}
+          1337-1452-666-4-v1.log.6                     # cached segment, legend:
+                                                       #    base_offset (e.g. 1337)
+                                                       #    committed_offset (e.g. 1452)
+                                                       #    size_bytes (e.g. 666)
+                                                       #    segment_term (e.g. 4)
+                                                       #    uploader_term (e.g. 6)
+          1337-1452-666-4-v1.log.6.tx                  # aborted transactions
+          1337-1452-666-4-v1.log.6.index               # index
+"""
+
+
+class JSONEncoder(json.JSONEncoder):
+    def default(self, obj):
+        return getattr(obj.__class__, "jsonable", lambda x: x)(obj)
+
+
+@dataclass
+class File:
+    path: str
+
+    def jsonable(self):
+        return {"type": "file", "path": self.path}
+
+
+@dataclass
+class Dir:
+    path: str
+
+    def jsonable(self):
+        return {"type": "dir", "path": self.path}
+
+
+@dataclass
+class DataDirMarker:
+    """ see storage_strict_data_init """
+    path: str
+
+    def jsonable(self):
+        return self.__dict__
+
+
+@dataclass
+class ConfigCache:
+    path: str
+
+    def jsonable(self):
+        return self.__dict__
+
+
+@dataclass
+class Accesstime:
+    path: str
+
+    def jsonable(self):
+        return self.__dict__
+
+
+@dataclass
+class CrashLoopFile:
+    path: str
+    size: int = -1
+
+    def jsonable(self):
+        return self.__dict__
+
+
+class CloudSegment:
```

Comment:
can we use `@dataclass` for all of those structures ?

Reply:
done, I hadn't known how to have mutable fields in `@dataclass` so used classes but looked up now and converted all to `@dataclass`